src/context.rs

```diff
@@ -326,59 +206,60 @@ impl<A: Actor> Context<A> {
+    ///         ctx.stop();
+    ///     }
+    /// }
+    ///
+    /// #[async_trait]
+    /// impl Handler<Joining> for MyActor {
+    ///     type Return = bool;
+    ///
+    ///     async fn handle(&mut self, _msg: Joining, ctx: &mut Context<Self>) -> bool {
+    ///         let addr = ctx.address().unwrap();
+    ///         let join = ctx.join(self, future::ready::<()>(()));
+    ///         let _ = addr.send(Stop).split_receiver().await;
+    ///
+    ///         // Actor is stopping, but the join should still evaluate correctly
+    ///         join.now_or_never().is_some()
+    ///     }
+    /// }
     ///
+    /// # #[cfg(feature = "with-smol-1")]
+    /// # smol::block_on(async {
+    /// let addr = MyActor.create(None).spawn(&mut xtra::spawn::Smol::Global);
+    /// assert!(addr.is_connected());
+    /// assert_eq!(addr.send(Joining).await, Ok(true)); // Assert that the join did evaluate the future
+    /// # })
     #[cfg_attr(docsrs, doc("```"))]
     #[cfg_attr(docsrs, doc(include = "../examples/interleaved_messages.rs"))]
     #[cfg_attr(docsrs, doc("```"))]
     pub async fn join<F, R>(&mut self, actor: &mut A, fut: F) -> R
     where
         F: Future<Output = R>,
     {
-        if !self.handle_self_notifications(actor).await {
-            self.stop();
-        }
-
         futures_util::pin_mut!(fut);
-
-        let addr_rx = self.receiver.clone();
-        let broadcast_rx = self.broadcast_receiver.clone();
-        let mut addr_recv = addr_rx.recv_async();
-        let mut broadcast_recv = broadcast_rx.recv_async();
-
-        while self.running == RunningState::Running {
-            let (next_msg, unfinished) = {
-                let next_msg = future::select(addr_recv, broadcast_recv);
-                futures_util::pin_mut!(next_msg);
-                match future::select(fut, next_msg).await {
-                    Either::Left((future_res, _)) => return future_res,
-                    Either::Right(tuple) => tuple,
-                }
-            };
-
-            let msg = match next_msg {
-                Either::Left((res, other)) => {
-                    broadcast_recv = other;
-                    addr_recv = addr_rx.recv_async();
-                    Either::Right(res.unwrap())
-                }
-                Either::Right((res, other)) => {
-                    addr_recv = other;
-                    broadcast_recv = broadcast_rx.recv_async();
-                    Either::Left(res.unwrap())
-                }
-            };
-
-            // To avoid broadcast starvation, try receive a broadcast here
-            if let Ok(Some(broadcast)) = broadcast_rx.try_recv() {
-                self.tick(Either::Left(broadcast), actor).await;
-            }
-
-            self.tick(msg, actor).await;
-            fut = unfinished;
+        match self.select(actor, fut).await {
+            Either::Left(res) => return res,
+            Either::Right(fut) => fut.await,
         }
```

Comment:
Both `join` and `select` are candidates to be moved to an extension crate, but for now it might be worth keeping for the symmetry of having `join` and `select` offered together, as they often are in most libraries

Reply:
Agreed.